pkg/cfn/builder/vpc_endpoint.go

```diff
@@ -0,0 +1,168 @@
+package builder
+
+import (
+	"fmt"
+	"strings"
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/service/ec2"
+	"github.com/aws/aws-sdk-go/service/ec2/ec2iface"
+
+	"github.com/pkg/errors"
+
+	api "github.com/weaveworks/eksctl/pkg/apis/eksctl.io/v1alpha5"
+
+	gfn "github.com/weaveworks/goformation/cloudformation"
+)
+
+// A VPCEndpointResourceSet represents the resources required for VPC endpoints
+type VPCEndpointResourceSet struct {
+	provider        provider
+	rs              *resourceSet
+	vpc             *gfn.Value
+	clusterConfig   *api.ClusterConfig
+	subnets         []subnetResource
+	clusterSharedSG *gfn.Value
+}
+
+type provider interface {
+	EC2() ec2iface.EC2API
+	Region() string
+}
+
+// NewVPCEndpointResourceSet creates a new VPCEndpointResourceSet
+func NewVPCEndpointResourceSet(provider provider, rs *resourceSet, clusterConfig *api.ClusterConfig, vpc *gfn.Value, subnets []subnetResource, clusterSharedSG *gfn.Value) *VPCEndpointResourceSet {
+	return &VPCEndpointResourceSet{
+		provider:        provider,
+		rs:              rs,
+		clusterConfig:   clusterConfig,
+		vpc:             vpc,
+		subnets:         subnets,
+		clusterSharedSG: clusterSharedSG,
+	}
+}
+
+// VPCEndpointServiceDetails holds the details for a VPC endpoint service
+type VPCEndpointServiceDetails struct {
+	ServiceName       string
+	Service           string
+	EndpointType      string
+	AvailabilityZones []string
+}
+
+// AddResources adds resources for VPC endpoints
+func (e *VPCEndpointResourceSet) AddResources() error {
+	endpointServices := append(api.RequiredEndpointServices(), e.clusterConfig.PrivateCluster.AdditionalEndpointServices...)
+	if e.clusterConfig.HasClusterCloudWatchLogging() && !e.hasEndpoint(api.EndpointServiceCloudWatch) {
+		endpointServices = append(endpointServices, api.EndpointServiceCloudWatch)
+	}
+	endpointServiceDetails, err := BuildVPCEndpointServices(e.provider.EC2(), e.provider.Region(), endpointServices)
+	if err != nil {
+		return errors.Wrap(err, "error building endpoint service details")
+	}
+
+	for _, endpointDetail := range endpointServiceDetails {
+		endpoint := &gfn.AWSEC2VPCEndpoint{
+			ServiceName:     gfn.NewString(endpointDetail.ServiceName),
+			VpcId:           e.vpc,
+			VpcEndpointType: gfn.NewString(endpointDetail.EndpointType),
+		}
+
+		if endpointDetail.EndpointType == ec2.VpcEndpointTypeGateway {
+			endpoint.RouteTableIds = e.routeTableIDs()
+		} else {
+			endpoint.SubnetIds = e.subnetsForAZs(endpointDetail.AvailabilityZones)
+			endpoint.PrivateDnsEnabled = gfn.NewBoolean(true)
+			endpoint.SecurityGroupIds = []*gfn.Value{e.clusterSharedSG}
+		}
+
+		resourceName := fmt.Sprintf("VPCEndpoint%s", strings.ToUpper(
+			strings.ReplaceAll(endpointDetail.Service, ".", ""),
+		))
+
+		// TODO attach policy document
```

Comment:
Reminder comment

Reply:
I'm planning to address this in a separate PR.